thunder/core/transform_common.py

```diff
@@ -33,6 +33,58 @@ def _remove_noop_subsymbols(bsym: BoundSymbol) -> None:
     bsym.subsymbols = nsbsyms
 
 
+def _inplace_copy_sanity_check(extrace: Trace):
+    """The sanity check is based on the sharp edge of nvfuser's `add_ouput(output, input)` interface,
+    it makes sure that the `copy_to` argument of `prims.copy_` is not used as input for any of its subsequent operators, except for the Return and Del operators
+
+    Anti-pattern:
+
+    .. code-block:: python
+
+        c = prims.copy_(a, b)
+        d = torch.add(b, b) # or d = torch.add(c, c)
+        return d
+
+    Do not use the `copy_to` variable `b` or `c` after it has been updated, use the `copy_from` variable `a` instead to reflect the dependency:
+
+    .. code-block:: python
+
+        c = prims.copy_(a, b)
+        d = torch.add(a, a)
+        return c
+    """
+    from thunder.core.trace import VariableInterface
+
+    inplace_copy_symbol_id = ("copy_", prims.PrimIDs.COPY_)
+    symbol_id_skip_list = (prims.PrimIDs.RETURN, prims.PrimIDs.DEL)
+    inplace_copy_to_arg: set[VariableInterface] = set()
```

Comment:
Do we want to keep a global `inplace_copy_to_arg` set?

I understand as the race only happens within a kernel, so the set should be reset when we check for a fusion symbol and clears afterwards?

Reply:
oh yes, thanks for pointing out this mistake. I misunderstood. The limitation of using `copy_to` is only inside one nvFusion. I'll fix it.